utils/estimated_histogram.hh

```diff
@@ -47,13 +47,302 @@
+     */
+    void clear() {
+        std::fill(_buckets.begin(), _buckets.end(), 0);
+    }
+
+    /*!
+     * \brief Add an item to the histogram
+     * Increments the count of the bucket closest to n, rounding DOWN.
+     */
+    void add(uint64_t n) {
+        _buckets.at(find_bucket_pos(n))++;
+    }
+
+    /*!
+     * \brief returns the smallest value that could have been added to this histogram
+     */
+    uint64_t min() const {
+        for (size_t i = 0; i < NumBuckets; i ++) {
+            if (_buckets[i] > 0) {
+                return get_bucket_lower_limit(i);
+            }
+        }
+        return 0;
+    }
+
+    /*!
+     * \brief returns the largest value that could have been added to this histogram.  If the histogram
+     * overflowed, returns UINT64_MAX.
+     */
+    uint64_t max() const {
+        if (_buckets[NumBuckets - 1] > 0) {
+            return UINT64_MAX;
+        }
+        for (int i = NumBuckets - 1; i >= 0; i--) {
+            if (_buckets[i] > 0) {
+                return get_bucket_upper_limit(i);
+            }
+        }
+        return 0;
+    }
+
+    /*!
+     * \brief merge a histogram to the current one.
+     */
+    approx_exponential_histogram& merge(const approx_exponential_histogram& b) {
+        for (size_t i = 0; i < NumBuckets; i++) {
+            _buckets[i] += b.get(i);
+        }
+        return *this;
+    }
+
+    template<uint64_t A, uint64_t B, size_t C>
+    friend approx_exponential_histogram<A, B, C> merge(approx_exponential_histogram<A, B, C> a, const approx_exponential_histogram<A, B, C>& b);
+
+    /*
+     * \brief returns the count in the given bucket
+     */
+    uint64_t get(size_t bucket) const {
+        return _buckets[bucket];
+    }
+
+    /*!
+     * \brief get a histogram quantile
+     *
+     * returns the estimated value at given quantile
+     */
+    uint64_t quantile(double quantile) const {
+        if (quantile < 0 || quantile > 1.0) {
+            throw std::runtime_error("Invalid quantile value " + std::to_string(quantile) + ". Value should be between 0 and 1");
+        }
+        auto c = count();
+
+        if (!c) {
+            return 0; // no data
+        }
+
+        auto pcount = uint64_t(std::floor(c * quantile));
+        uint64_t elements = 0;
+        for (size_t i = 0; i < NumBuckets - 2; i++) {
+            if (_buckets[i]) {
+                elements += _buckets[i];
+                if (elements >= pcount) {
+                    return get_bucket_lower_limit(i);
+                }
+            }
+        }
+        return Max; // overflowed value is in the requested quantile
+    }
+
+    /*!
+     * \brief returns the mean histogram value (average of bucket offsets, weighted by count)
+     */
+    uint64_t mean() const {
+        double elements = 0;
+        uint64_t sum = 0;
+        for (size_t i = 0; i < NumBuckets - 1; i++) {
+            elements += _buckets[i];
+            sum += _buckets[i] * get_bucket_lower_limit(i);
```

Comment:
sum can easily overflow uint64_t, no?

Reply:
theoretically, I can just switch to double all the way

Reply:
@amnonh 
Also, a running (and accurate) _sum can be maintained internally on each add()